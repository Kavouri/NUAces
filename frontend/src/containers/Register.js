import React from 'react';
import { request } from '../lib/requestWrapper';
import _ from 'lodash';

import Name from './registration/Name';
import { Birthday, MONTHS } from './registration/Birthday';
import Email from './registration/Email';
import Password from './registration/Password';
import ConfirmPassword from './registration/ConfirmPassword';
import Error from './registration/Error';

export default class Register extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      firstName: '',
      lastName: '',
      month: 'January',
      day: '',
      year: '',
      email: '',
      password: '',
      passwordConfirm: '',
      error: ''
    };

    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  handleChange(event) {
    this.setState({ [event.target.name]: event.target.value });
  }

  validateForm() {
    let error = '';
    _.forEachRight(this.state, (value, key) => {
      if (key !== 'error' && !this.state[key]) {
        error = `${key.split(/(?=[A-Z])/).map(word => _.upperFirst(word)).join(' ')} field is required`;
      }
    });

    if (!error) {

      if (this.state.day.concat(this.state.year).match(/[a-z]/i)) {
        error = 'DD-YYYY contains alphabetic characters.';
      }
      else if (this.state.day.length > 2 || this.state.year.length !== 4) {
        error = 'DD-YYYY has an invalid number of characters.';
      }
      else if (this.state.password !== this.state.passwordConfirm) {
        error = 'Passwords don\'t match';
      }
      else if (this.state.email.indexOf('@') < 0) {
        error = 'Email is not valid.';
      }
    }

    if (error) {
      this.setState({
        error
      });
    }
    return !error;
  }

  handleSubmit(event) {
    event.preventDefault();
    if (this.validateForm()) {
      const registrationForm = {
        name: `${this.state.firstName} ${this.state.lastName}`,
        dob: `${this.state.year}-${MONTHS[this.state.month]}-${this.state.day}`,
        email: this.state.email,
        password: this.state.password
      };

      request('/user', 'POST', registrationForm)
        .then((res) => {
            this.props.history.push('/');
          } else {
            this.setState({ error: res.body });
          }
        })
        .catch((error) => {
          this.setState({ error: error.error });
        });
    }
  }

  render() {
    return (
      <div className="registration">
      <form className="registration-form" onSubmit={this.handleSubmit}>
        <label>Name</label>
        <Name handleChange={this.handleChange} firstName={this.state.firstName} lastName={this.state.lastName}/>
        <label>Birthday</label>
        <Birthday handleChange={this.handleChange} month={this.state.month} day={this.state.day} year={this.state.year}/>
        <label>Email</label>
        <Email handleChange={this.handleChange} email={this.state.email}/>
        <label>Password</label>
        <Password handleChange={this.handleChange} password={this.state.password}/>
        <label>Confirm Password</label>
        <ConfirmPassword handleChange={this.handleChange} passwordConfirm={this.state.passwordConfirm}/>
        <button className="inner-registration register-button" size="3" type="submit">Register</button>
        {this.state.error && <Error error={this.state.error}/>}
      </form>
    </div>
    );
  }
}
