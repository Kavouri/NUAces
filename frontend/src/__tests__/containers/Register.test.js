import Register from '../../containers/Register';
import React from 'react';
import renderer from 'react-test-renderer';

describe('Register', () => {
  it('validates that a form is filled in', () => {
    const state = {
        firstName: '',
        lastName: '',
        email: '',
        password: '',
        passwordConfirm: '',
        month: 'January',
        day: '',
        year: '',  
        error: ''
    };

    let r = new Register();
    r.state = state;

    expect(() => {
        r.validateForm();
    }).toEqual(false);
  });

  it('validates that a form is filled in completely', () => {
    const state = {
        firstName: 'Austin',
        lastName: 'McKenna',
        email: 'mckenna.au@husky.neu.edu',
        password: 'hunter2',
        passwordConfirm: 'hunter2',
        month: 'January',
        day: '20',
        year: '1943',  
        error: ''
    };

    let r = new Register();
    r.state = state;

    expect(() => {
        r.validateForm();
    }).toEqual(true);
  });

  it('validates that a form is filled in completely(special-case)', () => {
    const state = {
        firstName: '',
        lastName: '',
        email: '',
        password: '',
        passwordConfirm: '',
        month: '',
        day: '',
        year: '',  
        error: 'ERROR!'
    };

    let r = new Register();
    r.state = state;

    expect(() => {
        !r.validateForm() && r.state.error;
    }).toEqual(true);
  });

  it('validates that the password form matches', () => {
    const state = {
        firstName: 'Austin',
        lastName: 'McKenna',
        email: 'mckenna.au@husky.neu.edu',
        password: 'hunter2',
        passwordConfirm: 'huntertwo',
        month: 'January',
        day: '20',
        year: '1943',  
        error: ''
    };

    let r = new Register();
    r.state = state;

    expect(() => {
        !r.validateForm() && r.state.error === 'Passwords don\'t match';
    }).toEqual(true);
  });

  it('validates that the email form is rational', () => {
    const state = {
        firstName: 'Austin',
        lastName: 'McKenna',
        email: 'mckenna.auatbingo.com!!!!BINGO!!!!husky.neu.edu',
        password: 'hunter2',
        passwordConfirm: 'hunter2',
        month: 'January',
        day: '20',
        year: '1943',  
        error: ''
    };

    let r = new Register();
    r.state = state;

    expect(() => {
        !r.validateForm() && r.state.error === 'Email is not valid';
    }).toEqual(true);
  });

  it('validates that the date form is stricly numeric', () => {
    const state = {
        firstName: 'Austin',
        lastName: 'McKenna',
        email: 'mckenna.au@husky.neu.edu',
        password: 'hunter2',
        passwordConfirm: 'hunter2',
        month: 'January',
        day: 'ha',
        year: 'nineteenfourtythree',  
        error: ''
    };

    let r = new Register();
    r.state = state;

    expect(() => {
        !r.validateForm() && r.state.error === 'DD-YYYY contains alphabetic characters.';
    }).toEqual(true);
  });

  it('validates that the date form has proper character count', () => {
    const state = {
        firstName: 'Austin',
        lastName: 'McKenna',
        email: 'mckenna.au@husky.neu.edu',
        password: 'hunter2',
        passwordConfirm: 'hunter2',
        month: 'January',
        day: '20',
        year: '20177',  
        error: ''
    };

    let r = new Register();
    r.state = state;

    expect(() => {
        !r.validateForm() && r.state.error === 'DD-YYYY has an invalid number of characters.';
    }).toEqual(true);
  });

  it('matches a Register snapshot', () => {
    const tree = renderer.create(<Register />).toJSON();
    expect(tree).toMatchSnapshot();
  });

});
