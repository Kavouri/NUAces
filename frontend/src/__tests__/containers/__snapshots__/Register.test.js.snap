// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register matches a Register snapshot 1`] = `
<div
  class="registration"
>
  <form
    class="registration-form"
    onSubmit={[Function]}
  >
    <label>
      Name
    </label>
    <div
      class="name-container inner-registration"
    >
      <input
        name="firstName"
        onChange={[Function]}
        placeholder="First"
        size="18"
        value=""
      />
      <input
        name="lastName"
        onChange={[Function]}
        placeholder="Last"
        size="18"
        value=""
      />
    </div>
    <label>
      Birthday
    </label>
    <div
      class="birthday-container inner-registration"
    >
      <select
        name="month"
        onChange={[Function]}
        value="January"
      >
        <option
          select="January"
        >
          January
        </option>
        <option
          select="February"
        >
          February
        </option>
        <option
          select="March"
        >
          March
        </option>
        <option
          select="April"
        >
          April
        </option>
        <option
          select="May"
        >
          May
        </option>
        <option
          select="June"
        >
          June
        </option>
        <option
          select="July"
        >
          July
        </option>
        <option
          select="August"
        >
          August
        </option>
        <option
          select="September"
        >
          September
        </option>
        <option
          select="October"
        >
          October
        </option>
        <option
          select="November"
        >
          November
        </option>
        <option
          select="December"
        >
          December
        </option>
      </select>
      <input
        name="day"
        onChange={[Function]}
        placeholder="Day"
        size="8"
        value=""
      />
      <input
        name="year"
        onChange={[Function]}
        placeholder="Year"
        size="16"
        value=""
      />
    </div>
    <label>
      Email
    </label>
    <div
      class="email-container inner-registration"
    >
      <input
        name="email"
        onChange={[Function]}
        placeholder="Email"
        size="24"
        type="email"
        value=""
      />
    </div>
    <label>
      Password
    </label>
    <div
      class="password-container inner-registration"
    >
      <input
        name="password"
        onChange={[Function]}
        placeholder="Password"
        size="24"
        type="password"
        value=""
      />
    </div>
    <label>
      Confirm Password
    </label>
    <div
      class="confirm-pw-container inner-registration"
    >
      <input
        name="passwordConfirm"
        onChange={[Function]}
        placeholder="Confirm Password"
        size="24"
        type="password"
        value=""
      />
    </div>
    <button
      class="inner-registration register-button"
      size="3"
      type="submit"
    >
      Register
    </button>
    
  </form>
</div>
`;
